add/refactor: Добавление скинов телескопичкам и рефактор изменения скинов для всех предметов

vovakons opened the pull request:
## Что этот ПР делает

1. Добавление новых скинов на телескопичку (*бронзовый, серебрянный, золотой. Все три варианта с свежими спрайтами от piotrthetchaikowsky, также варианты с тг*).
2. Рефактор кода связанного с применением скинов предметов (*обычно на альт клик*), перепишу на компонент для удобства.

## Почему это хорошо для игры

1. Новые скины телескопички.
2. Рефактор кода

## Демонстрация изменений

<details><summary>Демонстрации изменений</summary>
<p>

<img width="343" height="259" alt="image" src="https://github.com/user-attachments/assets/acef2c43-5055-4ea1-ab8e-2eb4a8526adf" />

<img width="338" height="337" alt="image" src="https://github.com/user-attachments/assets/1ca49587-3610-4114-9fce-e68ca22aa8aa" />

<img width="611" height="29" alt="image" src="https://github.com/user-attachments/assets/dc54eb33-2dc9-4d5f-a268-d73217b681d0" />


</p>
</details>

## Тестирование

Проверено на локалке.

vovakons committed:
batons icons move to new dmi file

Все спрайты батонов перемещены в новый dmi файл.

vovakons committed:
donation telescopic baton skins

Добавлена возможность поменять скин телескопички если есть тир донатера.
Тир 1 - бронзовая.
Тир 2 - серебрянная.
Тир 3 и 4 - золотая.

vovakons committed:
default telebaton skin update

Использование спрайта стандартной телескопички с тг вместо старой нашей.

vovakons committed:
classic variants for telebaton

Добавлены скины телескопичек от piotrthetchaikowsky.

vovakons committed:
item skins component

Компонент для рескина любого /obj/item.
Работает через альт клик.

vovakons committed:
gun reskins with new component

Использование нового компонента смены скина.
Удален старый код в obj/item/gun который отвечал за смену скина.

vovakons committed:
security baton skins

littleboobs commented on code/__DEFINES/misc.dm:
```diff
@@ -316,6 +316,16 @@
+// Donation level for Tier III
+#define DONATOR_TIER_III 3
+// Donation level for Tier IV
+#define DONATOR_TIER_IV 4
```
почему это в миске, а не в отдельном файле?

littleboobs commented on code/datums/components/_component.dm:
```diff
@@ -387,6 +387,12 @@
+	var/list/components = datum_components?[c_type]
+	if(!components)
+		return FALSE
+	return TRUE
```
зачем это...?

vovakons replied:
Проверка на существование компонента

dageavtobusnick replied:
Зачем? У тебя есть для этого геткомпонент

littleboobs commented on code/datums/components/item_skins.dm:
```diff
@@ -0,0 +1,91 @@
+
+	var/datum/item_skin_data/skin = skin_options[choice]
+	item.current_skin = skin.icon_state
+	to_chat(user, "Теперь [item.declent_ru(NOMINATIVE)] имеет облик '[skin.name]'. Познакомьтесь с новым дизайном.")
```
напиши нормальное предложение

vovakons replied:
я его скопировал откуда-то, предложи вариант лучше

littleboobs commented on code/game/objects/items.dm:
```diff
@@ -387,6 +390,9 @@ GLOBAL_DATUM_INIT(fire_overlay, /mutable_appearance, mutable_appearance('icons/g
 				. += span_notice("Обнаружено запечатанное заклинание \"[S.name]\" внутри.<br>")
 				break
 
+	if(ExistsComponent(/datum/component/item_skins))
```
💀

littleboobs commented on code/game/objects/items.dm:
```diff
@@ -387,6 +390,9 @@ GLOBAL_DATUM_INIT(fire_overlay, /mutable_appearance, mutable_appearance('icons/g
 				break
 
+	if(ExistsComponent(/datum/component/item_skins))
+		. += span_notice("Используйте Alt-click чтобы выбрать скин.")
```
```suggestion
		. += span_notice("Используйте <b>Alt+ЛКМ</b>, чтобы выбрать скин.")
```

dageavtobusnick commented on code/modules/projectiles/guns/projectile/revolver.dm:
```diff
@@ -145,18 +145,20 @@
+		new /datum/item_skin_data(name = "White Gold", icon_state = "detective_gold"),
+		new /datum/item_skin_data(name = "Gold Wood", icon_state = "detective_gold_alt"),
+		new /datum/item_skin_data(name = "The Peacemaker", icon_state = "detective_peacemaker"),
+		new /datum/item_skin_data(name = "Silver", icon_state = "detective_silver"),
```
ммм. А нам точно нужно для каждого объекта плодить дохулиард датумов?

dageavtobusnick commented on code/datums/components/item_skins.dm:
```diff
@@ -0,0 +1,91 @@
+ * Add skins feature to /obj/item by alt_click.
+ */
+/datum/component/item_skins
+	var/list/skins
```
Почему не элемент? Я не вижу тут состояний, которые нужно хранить в компоненте.

dageavtobusnick commented on code/datums/components/item_skins.dm:
```diff
@@ -0,0 +1,91 @@
+	/// Icon for radial menu (if null - use icon_state)
+	var/menu_icon_state = null
+	/// Minimal donater tier (0 for allow all players)
+	var/donation_tier = 0
```
Посмотри как сделаны скины для боргов. Тебе по сути нужно сделать подобное в плане количества датумов. 1 скин 1 тип датума и 1 объект.

dageavtobusnick reviewed (changes requested):
Нужно переделать

vovakons committed:
donation defines to new dm file

Создан новый файл donat.dm для констант доната.

items skins component to element

Перевод в элемент вместо компонента.
Удален лишний прок ExistsComponent, вместо него добавлен флаг для показа
подсказки в описании если есть возможность смены скина.

item skins to global datums list

Создан глобальный список датумов для скинов.
Немного переработана логика скинов в элементе.

duplicate skin definiition fix

fix radial check logic

skin apply text fix

fix register element

change gun.dm for test

vovakons committed:
Merge branch 'master220' into telebaton-skins

vovakons committed:
fix after merge

littleboobs commented on code/datums/elements/item_skins.dm:
```diff
@@ -0,0 +1,72 @@
+
+	var/datum/item_skin_data/skin = skin_options[choice]
+	item.current_skin = skin.icon_state
+	to_chat(user, "На [item.declent_ru(ACCUSATIVE)] установлен скин '[choice]'.")
```
```suggestion
	to_chat(user, "На [item.declent_ru(ACCUSATIVE)] установлен скин \"[choice]\".")
```

littleboobs commented on code/modules/projectiles/gun.dm:
```diff
@@ -657,12 +630,9 @@
 		return NONE
 	if(user.incapacitated() || HAS_TRAIT(user, TRAIT_HANDS_BLOCKED))
-		to_chat(user, span_warning("Вы не можете сделать это сейчас!"))
+		to_chat(user, span_warning("Вы не можете сделать это сейчас"))
```
```suggestion
		to_chat(user, span_warning("Вы не можете сделать это сейчас!"))
```

littleboobs commented on paradise.dme:
```diff
@@ -792,6 +794,7 @@
 #include "code\datums\ruins\bridges.dm"
 #include "code\datums\ruins\lavaland.dm"
 #include "code\datums\ruins\space.dm"
+#include "code\datums\skins\skins.dm"
```
не понятно про какие скины речь, уточни, что это донат скины

vovakons commented:
как же локализатор душит 💀

vovakons committed:
review fixes

vovakons committed:
Merge branch 'master220' into telebaton-skins

PlayerUnknown14 commented on code/datums/elements/item_skins.dm:
```diff
@@ -0,0 +1,72 @@
+		skin_options[skin.name] = image(icon = (skin.icon ? skin.icon : item.icon), icon_state = (skin.menu_icon_state ? skin.menu_icon_state : skin.icon_state) )
+
+	if(!length(skin_options))
+		to_chat(user, span_warning("Для получения скинов необходимо сделать пожертвование в Discord сообществе."))
```
```suggestion
		to_chat(user, span_warning("Для получения скинов необходимо сделать пожертвование в Discord-сообществе проекта!"))
```

PlayerUnknown14 commented on code/datums/skins/donat_skins.dm:
Остальные скины не хочешь заодно перевести? Не настаиваю, но было бы неплохо.

vovakons replied:
остальные скины это ты про щитспавн скины боргов? или про названия скинов для энфорсера и прочих?

PlayerUnknown14 replied:
Да. Все, что в файле.

vovakons replied:
а можно не надо? 💀

vovakons replied:
могу сделать, если не будешь сильно душить

PlayerUnknown14 replied:
> могу сделать, если не будешь сильно душить

Не буду. Жду перевод.

vovakons committed:
localization fix

vovakons committed:
donat skins localization try 1

vovakons committed:
donat skins localization try 2

PlayerUnknown14 commented on code/datums/skins/donat_skins.dm:
```diff
@@ -0,0 +1,263 @@
+	icon_state = "specter_red"
+
+/datum/item_skin_data/specter/green_slide
+	name = "Зеленый затвор"
```
https://github.com/ss220-space/Paradise/blob/master220/.github/guides/LOCALIZATION.md#буква-ё

PlayerUnknown14 commented on code/datums/skins/donat_skins.dm:
```diff
@@ -0,0 +1,263 @@
+	icon_state = "specter_tan"
+
+/datum/item_skin_data/specter/green_handle
+	name = "Зеленый корпус"
```
Почему корпус, если это рукоять?

PlayerUnknown14 commented on code/datums/skins/donat_skins.dm:
```diff
@@ -0,0 +1,263 @@
+	icon_state = "enforcer_green"
+
+/datum/item_skin_data/enforcer/tan_slide
+	name = "Песочный затвор"
```
Так бежевый или песочный?

PlayerUnknown14 commented on code/datums/skins/donat_skins.dm:
```diff
@@ -0,0 +1,263 @@
+	icon_state = "detective_leopard"
+
+/datum/item_skin_data/detective_revolver/black_panther
+	name = "Черна пантера"
```
ё я

PlayerUnknown14 commented on code/datums/skins/donat_skins.dm:
```diff
@@ -0,0 +1,263 @@
+	icon_state = "detective_gold"
+
+/datum/item_skin_data/detective_revolver/gold_wood
+	name = "Золотое дерево"
```
```suggestion
	name = "Позолота"
```

PlayerUnknown14 commented on code/datums/skins/donat_skins.dm:
```diff
@@ -0,0 +1,263 @@
+	icon_state = "dshotgun"
+
+/datum/item_skin_data/doublebarrel/dark_red
+	name = "Темно-красная отделка"
```
Тёмно-красный

PlayerUnknown14 commented on code/datums/skins/donat_skins.dm:
```diff
@@ -0,0 +1,263 @@
+	icon_state = "dshotgun-l"
+
+/datum/item_skin_data/doublebarrel/rosewood
+	name = "Палисандр"
```
овый

vovakons committed:
donat skins localization try 3

PlayerUnknown14 commented on code/datums/skins/donat_skins.dm:
```diff
@@ -107,23 +107,23 @@
 
 /datum/item_skin_data/specter/green_handle
-	name = "Зеленый корпус"
+	name = "Зелёный рукоять"
```
ая

PlayerUnknown14 commented on code/datums/skins/donat_skins.dm:
```diff
@@ -215,15 +215,15 @@
 
 /datum/item_skin_data/detective_revolver/black_panther
-	name = "Черна пантера"
+	name = "Черная пантера"
```
ё

vovakons committed:
donat skins localization try 4

PlayerUnknown14 commented on code/datums/skins/donat_skins.dm:
```diff
@@ -0,0 +1,263 @@
+	donation_tier = 3
+
+/datum/item_skin_data/telescopic_baton/gold_classic
+	name = "Классическая золотая телескопическая дубинка"
```
"телескопическая дубинка" убери у всех.

PlayerUnknown14 reviewed (changes requested):
супер ультра мега ласт

vovakons committed:
donat skins localization try 5
